### PR TITLE
Memory savings by reusing the same buffer across layers in backward

### PR DESCRIPTION
**Pull request by ngc92**
reuse memory buffers across layers during the backward pass

### CODE REVIEW

File: train_gpt2.cu [hunk 1]
```diff
@@ -601,12 +601,12 @@ __global__ void matmul_backward_bias_kernel_faster(float* dbias, const float* do
     }
     // write the final result (at thread 0) to global memory
     if (tid == 0) {
-        dbias[o] = shared[0];
+        dbias[o] += shared[0];
```
Comment (karpathy) [inline train_gpt2.cu:604]: every change is += -> =, except here = -> += ?

File: train_gpt2.cu [hunk 2]
```diff
@@ -1533,23 +1567,25 @@ void gpt2_backward(GPT2 *model) {
...
-        float* dl_att = grads_acts.att + l * B * NH * T * T;
-        float* dl_v_accum = grads_acts.v_accum + l * B * T * C;
-        float* dl_attproj = grads_acts.attproj + l * B * T * C;
-        float* dl_residual2 = grads_acts.residual2 + l * B * T * C;
-        float* dl_ln2 = grads_acts.ln2 + l * B * T * C;
-        float* dl_fch = grads_acts.fch + l * B * T * 4*C;
-        float* dl_fch_gelu = grads_acts.fch_gelu + l * B * T * 4*C;
-        float* dl_fcproj = grads_acts.fcproj + l * B * T * C;
-        float* dl_residual3 = grads_acts.residual3 + l * B * T * C;
+        float* dl_ln1 = grads_acts.ln1;
```
Comment (karpathy) [inline train_gpt2.cu:1570]: wait if we're only doing it like this layer by layer, aren't we able to allocate much less for `grad_acts`? We don't need the multiplication by `L`?

File: train_gpt2.cu [hunk 3]
```diff
@@ -1458,8 +1469,31 @@ void gpt2_backward(GPT2 *model) {
...
+        // allocate buffers for weight gradients
         model->grads_memory = malloc_and_point_parameters(&model->grads, model->param_sizes, 1);
-        model->grads_acts_memory = malloc_and_point_activations(&model->grads_acts, model->act_sizes);
+
+        // determine buffer size for activation gradients
+        // _mostly_ allocate as if we had just a single layer,
+        // because we're going to reuse memory
+        size_t bw_act_sizes[NUM_ACTIVATION_TENSORS];
+        GPT2Config cfg = model->config;
+        cfg.num_layers = 1;
```
Comment (karpathy) [inline train_gpt2.cu:1480]: ah, I just noticed this critical part. L = 1

File: train_gpt2.cu (hunk 1 above)
Comment (ngc92) [reply train_gpt2.cu:604]: this is the bias kernel, right. So if we're doing gradient accumulation over multiple steps, we would want to add these up, right?
Technically, this isn't part of the memory changes, but when checking each kernel, I noticed this.

Comment (karpathy) [reply train_gpt2.cu:604]: But aren't we able to keep it as = like all the other kernels?

**Comment by karpathy** [comment]
The danger here is that we have two forking points in our graph:

- residual layers create a fork
- `wte` has a fork - the same tensor is used both for token embeddings AND for final matmul

we have to make sure these two parts do a += correctly. I'm trying to make sure this happens in the code but it's not obvious yet.

**Comment by karpathy** [comment]
1) ok `wte` will be ok because:
- matmul now does =
- but encoder backward still does += 
OK

**Pull request merged**